server/app/services/applicant/ApplicantData.java

```diff
@@ -114,4 +118,25 @@ public void setUserName(
       putString(WellKnownPaths.APPLICANT_LAST_NAME, lastName);
     }
   }
+
+  @Override
+  public String asJsonString() {
+    if (!getFailedUpdates().isEmpty()) {
+      throw new IllegalStateException("data cannot be serialized since there were failed updates");
+    }
+    return super.asJsonString();
+  }
+
+  public void setFailedUpdates(ImmutableMap<Path, String> updates) {
```

Comment:
nit: Add a comment? It may not be clear to others that the values in the map are the inputs that caused the "failed updates"

Reply:
Done!